### PR TITLE
Move initialization of YieldProcessorNormalized to the finalizer thread

### PR DESCRIPTION
**Pull request by kouvel**
Fixes https://github.com/dotnet/coreclr/issues/13984
- Also moved relevant functions out of the Thread class as requested in the issue
- For some reason, after moving the functions out of the Thread class, YieldProcessorNormalized was not getting inlined anymore. It seems to be important to have it be inlined such that the memory loads are hoisted out of outer loops. To remove the dependency on the compiler to do it (even with forceinline it's not possible to hoist sometimes, for instance InterlockedCompareExchnage loops), changed the signatures to do what is intended.

### CODE REVIEW

File: src/vm/threads.cpp [hunk 1]
```diff
@@ -11747,25 +11749,29 @@ ULONGLONG Thread::QueryThreadProcessorUsage()
...
 
-CrstStatic Thread::s_initializeYieldProcessorNormalizedCrst;
-int Thread::s_yieldsPerNormalizedYield = 0;
-int Thread::s_optimalMaxNormalizedYieldsPerSpinIteration = 0;
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// YieldProcessorNormalized
+
+// Defaults are for when InitializeYieldProcessorNormalized has not yet been called or when no measurement is done, and are
+// tuned for Skylake processors
+int g_yieldsPerNormalizedYield = 1; // 9 for pre-Skylake
```
Comment (stephentoub) [inline src/vm/threads.cpp]: What does this "9 for pre-Skylake" comment mean?

File: src/vm/finalizerthread.cpp [hunk 2]
```diff
@@ -744,6 +744,8 @@ DWORD WINAPI FinalizerThread::FinalizerThreadStart(void *args)
 #endif
             GetFinalizerThread()->SetBackground(TRUE);
 
+            EnsureYieldProcessorNormalizedInitialized();
```
Comment (stephentoub) [inline src/vm/finalizerthread.cpp]: Am I understanding the change correctly that until the finalizer thread gets to this point, spinning will be done using a default value, and then once this runs, that value will be updated appropriately based on the system?  I think that makes sense, just wanted to confirm.

**Comment by kouvel** [comment]
I just realized that the scheme doesn't work for the back-off version, I'll change it around a bit

**Comment by kouvel** [comment]
Going to revoke your approval as a result

File: src/vm/threads.cpp (hunk 1 above)
Comment (kouvel) [reply src/vm/threads.cpp]: For pre-Skylake processors it would be about 9 yields per normalized yield

**Comment by stephentoub** [comment]
Thanks

File: src/vm/finalizerthread.cpp (hunk 2 above)
Comment (kouvel) [reply src/vm/finalizerthread.cpp]: Yes that's right

File: src/vm/threads.cpp (hunk 1 above)
Comment (stephentoub) [reply src/vm/threads.cpp]: Ok.  It'd be helpful to expand the comment to highlight that this is an estimate for what the value will be, not that we're somewhere setting this to a hardcoded "9".

Comment (kouvel) [reply src/vm/threads.cpp]: Done

**Comment by tarekgh** [comment]
>     return optimalMaxNormalizedYieldsPerSpinIteration;

any reason why we cannot just return g_optimalMaxNormalizedYieldsPerSpinIteration?

---
Refers to: src/vm/comsynchronizable.cpp:1641 in ffab10f. [](commit_id = ffab10fb6621a0427645bcf3496d41ed2fc6873a, deletion_comment = False)

**Comment by kouvel** [comment]
> any reason why we cannot just return g_optimalMaxNormalizedYieldsPerSpinIteration?

I think it's ok to do that for this case, the other QCALL functions are organized in this way I think to make sure all of the time-consuming work happens between BEGIN_QCALL and END_QCALL so I just followed the same pattern.

**Comment by kouvel** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

**Pull request merged**